pkg/epp/requestcontrol/director.go

```diff
@@ -0,0 +1,177 @@
+	"sigs.k8s.io/gateway-api-inference-extension/api/v1alpha2"
+	"sigs.k8s.io/gateway-api-inference-extension/pkg/epp/backend"
+	"sigs.k8s.io/gateway-api-inference-extension/pkg/epp/datastore"
+	"sigs.k8s.io/gateway-api-inference-extension/pkg/epp/handlers"
+	schedulingtypes "sigs.k8s.io/gateway-api-inference-extension/pkg/epp/scheduling/types"
+	errutil "sigs.k8s.io/gateway-api-inference-extension/pkg/epp/util/error"
+	logutil "sigs.k8s.io/gateway-api-inference-extension/pkg/epp/util/logging"
+)
+
+type Scheduler interface {
+	Schedule(ctx context.Context, b *schedulingtypes.LLMRequest) (result *schedulingtypes.Result, err error)
+}
+
+type Director struct {
+	datastore datastore.Datastore
+	scheduler Scheduler
+}
+
+func NewDirector(datastore datastore.Datastore, scheduler Scheduler) *Director {
+	return &Director{
+		datastore: datastore,
+		scheduler: scheduler,
+	}
+}
+
+// HandleRequest always returns the requestContext even in the error case, as the request context is used in error handling.
+func (d *Director) HandleRequest(ctx context.Context, reqCtx *handlers.RequestContext) (*handlers.RequestContext, error) {
+	logger := log.FromContext(ctx)
+
+	// Resolve target models.
+	var ok bool
+	requestBodyMap := reqCtx.Request.Body
+	reqCtx.Model, ok = requestBodyMap["model"].(string)
+	if !ok {
+		return reqCtx, errutil.Error{Code: errutil.BadRequest, Msg: "model not found in request"}
+	}
+	prompt, ok := requestBodyMap["prompt"].(string)
+	if !ok {
+		return reqCtx, errutil.Error{Code: errutil.BadRequest, Msg: "prompt not found in request"}
+	}
+
+	// NOTE: The nil checking for the modelObject means that we DO allow passthrough currently.
+	// This might be a security risk in the future where adapters not registered in the InferenceModel
+	// are able to be requested by using their distinct name.
+	modelObj := d.datastore.ModelGet(reqCtx.Model)
+	if modelObj == nil {
+		return reqCtx, errutil.Error{Code: errutil.BadConfiguration, Msg: fmt.Sprintf("error finding a model object in InferenceModel for input %v", reqCtx.Model)}
+	}
+
+	if len(modelObj.Spec.TargetModels) > 0 {
+		reqCtx.ResolvedTargetModel = RandomWeightedDraw(logger, modelObj, 0)
+		if reqCtx.ResolvedTargetModel == "" {
+			return reqCtx, errutil.Error{Code: errutil.BadConfiguration, Msg: fmt.Sprintf("error getting target model name for model %v", modelObj.Name)}
+		}
+	}
+
+	llmReq := &schedulingtypes.LLMRequest{
+		Model:               reqCtx.Model,
+		ResolvedTargetModel: reqCtx.ResolvedTargetModel,
+		Critical:            modelObj.Spec.Criticality != nil && *modelObj.Spec.Criticality == v1alpha2.Critical,
+		Prompt:              prompt,
+		Headers:             reqCtx.Request.Headers,
+	}
+	logger.V(logutil.DEBUG).Info("LLM request assembled", "request", llmReq)
+	results, err := d.Dispatch(ctx, llmReq)
+
+	// Insert target endpoint to instruct Envoy to route requests to the specified target pod.
+	// Attach the port number
+	reqCtx, err = d.PostDispatch(ctx, reqCtx, results)
```

Comment:
and if you agree on previous comments, this is not `PostDispatch` (cause no dispatching was done here).
this is more of `handleSchedulerResults` function.